File: z-функция/step4/C.java
package com.company.codeforces.step4;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.Arrays;

public class C {
    private static int countOfQueries;

    public static void main(String[] args) throws IOException {
        enterInput(System.in);
    }

    private static void enterInput(InputStream in) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(in));
        countOfQueries = Integer.parseInt(bf.readLine());
        for (int i = 0; i < countOfQueries; i++) {
            String current = bf.readLine();
            findMinPeriod(current);
        }
    }

    private static void findMinPeriod(String currentString) {
        currentString+=currentString;
        int[] z = new int[currentString.length()];
        int[] mass = new int[currentString.length()];

        for (int i = 1, l = 0, r = 0; i < z.length; i++) {
            if (i <= r) {
                z[i] = Math.min(z[i - l], r - i + 1);
            }
            while (z[i] + i < currentString.length()
                    && currentString.charAt(z[i]) == currentString.charAt(i + z[i]))
                ++z[i];

            if (r < i + z[i] - 1) {
                l = i;
                r = i + z[i] - 1;
            }

//            for(int j=1; j<=z[i]; j++){
//                mass[j-1]++;
//            }
            if (z[i] > 0) {
                ++mass[z[i] - 1];
            }
        }
        int counter = 1;
        System.out.println(Arrays.toString(z));
        System.out.println(Arrays.toString(mass));
        for (int i = mass.length - 1; i >= 0; i--) {
            z[i] = mass[i] + counter;
            if (mass[i] != 0) {
                counter += mass[i];
            }
        }
        for (int i = 0; i < z.length; i++) {
            System.out.print(z[i] + " ");
        }
        System.out.println();
    }
}

File: z-функция/step4/C.cpp
// Example program
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

void z(string s) {
	vector<int> z(s.length());
	vector<int> mass(s.length());
	int n = s.length();
	for (int i = 1, l = 0, r = 0; i < n; ++i) {
		if (i <= r)
			z[i] = min(r - i + 1, z[i - l]);
		while (i + z[i] < n && s[z[i]] == s[i + z[i]])
			++z[i];
		if (i + z[i] - 1 > r) {
			l = i;
			r = i + z[i] - 1;
		}
		if (z[i] > 0) {
			++mass[z[i] - 1];
		}

	}
	int counter = 1;
	for (int i = mass.size() - 1; i >= 0; i--) {
		z[i] = mass[i] + counter;
		if (mass[i] != 0) {
			counter += mass[i];
		}
	}
	for (int i = 0; i < n; i++) {
		std::cout << z[i] << " ";
	}

}

int main()
{
	std::cout.sync_with_stdio(false);
	std::cin.sync_with_stdio(false);
	int count;
	std::cin >> count;
	for (int i = 0; i < count; i++) {
		string str;
		std::cin >> str;
		z(str);
	}
}
